### ISSUE TITLE
Exception caught by widgets library... Output says to report a bug...

### ISSUE THREAD

**Issue by dspells** [flutter/flutter#75025]
## Steps to Reproduce

1. I'm having trouble reproducing this. The debug output said to report it. I'll to reproduce this in verbose mode and update the bug when I can. In the mean time I hope the Exception output is helpful.

**Expected results:** 

**Actual results:** 

<details>
  <summary>Logs</summary>

```
E/flutter (15255): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: RangeError (index): Invalid value: Valid value range is empty: -1
E/flutter (15255): #0      List.[] (dart:core-patch/growable_array.dart:254:60)
E/flutter (15255): #1      NavigatorState._flushHistoryUpdates (package:flutter/src/widgets/navigator.dart:3796:34)
E/flutter (15255): #2      NavigatorState.pop (package:flutter/src/widgets/navigator.dart:4917:7)
E/flutter (15255): #3      NavigatorState.popUntil (package:flutter/src/widgets/navigator.dart:4949:7)
E/flutter (15255): #4      Navigator.popUntil (package:flutter/src/widgets/navigator.dart:2600:27)
E/flutter (15255): #5      _DiscoveryScreenState._bluetoothConnect.<anonymous closure>.<anonymous closure> (package:intellaview_mobile/screens/discovery.dart:223:25)
E/flutter (15255): #6      _rootRunUnary (dart:async/zone.dart:1194:47)
E/flutter (15255): #7      _CustomZone.runUnary (dart:async/zone.dart:1097:19)
E/flutter (15255): #8      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1002:7)
E/flutter (15255): #9      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
E/flutter (15255): #10     _DelayedData.perform (dart:async/stream_impl.dart:611:14)
E/flutter (15255): #11     _StreamImplEvents.handleNext (dart:async/stream_impl.dart:730:11)
E/flutter (15255): #12     _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:687:7)
E/flutter (15255): #13     _rootRun (dart:async/zone.dart:1178:47)
E/flutter (15255): #14     _CustomZone.run (dart:async/zone.dart:1090:19)
E/flutter (15255): #15     _CustomZone.runGuarded (dart:async/zone.dart:994:7)
E/flutter (15255): #16     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1034:23)
E/flutter (15255): #17     _rootRun (dart:async/zone.dart:1186:13)
E/flutter (15255): #18     _CustomZone.run (dart:async/zone.dart:1090:19)
E/flutter (15255): #19     _CustomZone.runGuarded (dart:async/zone.dart:994:7)
E/flutter (15255): #20     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1034:23)
E/flutter (15255): #21     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
E/flutter (15255): #22     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
E/flutter (15255): 

======== Exception caught by widgets library =======================================================
The following assertion was thrown building Navigator-[GlobalObjectKey<NavigatorState> _WidgetsAppState#d6540](dirty, dependencies: [UnmanagedRestorationScope, HeroControllerScope], state: NavigatorState#f39a5(tickers: tracking 5 tickers)):
'package:flutter/src/widgets/navigator.dart': Failed assertion: line 5142 pos 12: '!_debugLocked': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md

The relevant error-causing widget was: 
  MaterialApp file:///Users/david.spells/Documents/Code/svn/apconMobile/trunk/flutter/intellaview_mobile/lib/main.dart:265:12
When the exception was thrown, this was the stack: 
#2      NavigatorState.build (package:flutter/src/widgets/navigator.dart:5142:12)
#3      StatefulElement.build (package:flutter/src/widgets/framework.dart:4802:27)
#4      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4685:15)
#5      StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4857:11)
#6      Element.rebuild (package:flutter/src/widgets/framework.dart:4379:5)
...
====================================================================================================

```

```
```

[✓] Flutter (Channel beta, 1.25.0-8.3.pre, on macOS 11.1 20C69 darwin-x64, locale en)
    • Flutter version 1.25.0-8.3.pre at /Users/david.spells/flutter
    • Framework revision 5d36f2e7f5 (2 weeks ago), 2021-01-14 15:57:49 -0800
    • Engine revision 7a8f8ca02c
    • Dart version 2.12.0 (build 2.12.0-133.7.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/david.spells/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/david.spells/Library/Android/sdk
    • Java binary at: /Users/david.spells/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/david.spells/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.3.2)
    • IntelliJ at /Users/david.spells/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 52.2.5
    • Dart plugin version 203.6912

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (1 available)
    • Pixel 4a (mobile) • 08071JEC202890 • android-arm64 • Android 11 (API 30)

• No issues found!
</details>

**Comment by iapicca** [comment]
@dspells 

> ======== Exception caught by widgets library =======================================================
> The following assertion was thrown building Navigator-[GlobalObjectKey<NavigatorState> _WidgetsAppState#d6540](dirty, dependencies: [UnmanagedRestorationScope, HeroControllerScope], state: NavigatorState#f39a5(tickers: tracking 5 tickers)):
> 'package:flutter/src/widgets/navigator.dart': Failed assertion: line 5142 pos 12: '!_debugLocked': is not true.

this might  be an error in your code, not necessarily a bug with flutter
please provide a [minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)

**Comment by TahaTesser** [comment]
Hi @dspells 
Can you please provide a complete reproducible minimal code sample
Thank you

**Comment by dspells** [comment]
Unfortunately this is not easily reproducible. It only happens when bluetooth on the phone "randomly" disconnects. Perhaps you could give me a clue as to why Navigator.popUntil would throw an exception. It says !_debugLocked is not true but I don't know why _debugLocked would be true.

Here is the gist of what is happening though.

- I connect to a BLE device
- I start listening for "disconnects"
- I push another screen onto the stack
- For some reason bluetooth disconnects and calls my listener
- In the listener I call Navigator.popUntil(context, (route) => false); in an attempt to get back to the device discovery screen

The error text says
`Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.`
Is there any more information that you can provide to help me determine why the code is failing and how to avoid calling Navigator.popUntil while _debugLocked is true.

**Comment by iapicca** [comment]
> [...] It only happens when bluetooth on the phone "randomly" disconnects. 
> [...]
> * I connect to a BLE device
> * I start listening for "disconnects"
> * I push another screen onto the stack
> * For some reason bluetooth disconnects and calls my listener

@dspells 
this seems to involve 3rd party plugins
for the issue to be workable need to be reproducible
please provide a [minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)
**without 3rd party plugins**

**Comment by dspells** [comment]
It sounds like you don't know when and why _debugLocked is set to true.

**Comment by dspells** [comment]
After looking through the Navigator code I came to the conclusion that if you call popUntil when there is nothing to pop _flushHistoryUpdates will crash. I added a line to check for if (Navigator.canPop before calling popUntil). I hope this fixes the problem. My hypothesis is that my listener was receiving a message before pushing and new screen on the stack. I don't know if this information is useful to you or not.
